rita_client/src/exit_manager/mod.rs

```diff
@@ -72,78 +76,121 @@ pub struct ExitBlacklist {
     /// Hashmap of Structs containing information of current exit we are connected to and tracking exit, if connected to one
     /// It also holds information about metrics and degradation values. Look at doc comment on 'set_best_exit' for more
     /// information on what these mean
     pub selected_exit_list: HashMap<String, SelectedExit>,
+    /// This struct hold infomation about exits that have misbehaved and are blacklisted, or are being watched
+    /// to being blacklisted through bad responses.
+    pub exit_blacklist: ExitBlacklist,
+}
+
+/// An actor which pays the exit
+#[derive(Default, Clone)]
+pub struct ExitManager {
+    pub nat_setup: bool,
     /// Every tick we query an exit endpoint to get a list of exits in that cluster. We use this list for exit switching
     pub exit_list: ExitList,
+    /// Store last exit here, when we see an exit change, we reset wg tunnels
+    pub last_exit: Option<IpAddr>,
+}
+
+pub fn set_last_exit(e: Option<IpAddr>, em_state: &mut ExitManager) {
+    em_state.last_exit = e;
+}
+
+pub fn get_last_exit(em_state: ExitManager) -> Option<IpAddr> {
+    em_state.last_exit
 }
 
 /// This functions sets the exit list ONLY IF the list arguments provived is not empty. This is need for the following edge case:
 /// When an exit goes down, the endpoint wont repsond, so we have no exits to switch to. By setting only when we have a length > 1
 /// we assure that we switch when an exit goes down
-pub fn set_exit_list(list: ExitList) -> bool {
+pub fn set_exit_list(list: ExitList, em_state: &mut ExitManager) -> bool {
     if !list.exit_list.is_empty() {
-        EXIT_MANAGER.write().unwrap().exit_list = list;
+        em_state.exit_list = list;
         return true;
     }
     false
 }
 
-pub fn get_exit_list() -> ExitList {
-    EXIT_MANAGER.read().unwrap().exit_list.clone()
+pub fn get_exit_list(em_state: ExitManager) -> ExitList {
+    em_state.exit_list
 }
 
 pub fn get_selected_exit(exit: String) -> Option<IpAddr> {
-    match EXIT_MANAGER.read().unwrap().selected_exit_list.get(&exit) {
+    match SELECTED_EXIT_LIST
+        .write()
+        .unwrap()
+        .selected_exit_list
+        .get(&exit)
+    {
         Some(a) => a.selected_id,
         None => None,
     }
 }
 
 pub fn get_selected_exit_metric(exit: String) -> Option<u16> {
-    match EXIT_MANAGER.read().unwrap().selected_exit_list.get(&exit) {
+    match SELECTED_EXIT_LIST
+        .read()
+        .unwrap()
+        .selected_exit_list
+        .get(&exit)
+    {
         Some(a) => a.selected_id_metric,
         None => None,
     }
 }
 
 pub fn get_selected_exit_tracking(exit: String) -> Option<IpAddr> {
-    match EXIT_MANAGER.read().unwrap().selected_exit_list.get(&exit) {
+    match SELECTED_EXIT_LIST
+        .read()
+        .unwrap()
+        .selected_exit_list
+        .get(&exit)
+    {
         Some(a) => a.tracking_exit,
         None => None,
     }
 }
 
 pub fn get_selected_exit_degradation(exit: String) -> Option<u16> {
```

Comment:
why not combine the above three functions and just return the whole exit struct?

Reply:
fixed